Delete node causing scheduler crash

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


**What happened**:
If we delete node first, then remove pods on this node, the scheduler will crash in `RemovePod` function

**What you expected to happen**:
Not crashed

**How to reproduce it (as minimally and precisely as possible)**:
1. Create some pods on node A
2. kubectl delete node A
3. kubectl delete pods on node A

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.16-1.18, even master
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

> I patched the unit test shown in #89908 into the `release-1.18`, `release-1.17` and `release-1.16` branches. None crashed.
> I'll await for the crash dump.

Here is the fresh dump.

```
E0407 00:04:49.430956   15014 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 247 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x188b100, 0x2c28480)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
panic(0x188b100, 0x2c28480)
	/usr/local/go/src/runtime/panic.go:967 +0x166
k8s.io/kubernetes/pkg/scheduler/types.(*NodeInfo).RemovePod(0xc0005f1450, 0xc00056e800, 0xc000884930, 0x9)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/types/node_info.go:555 +0x610
k8s.io/kubernetes/pkg/scheduler/internal/cache.(*schedulerCache).removePod(0xc00034e360, 0xc00056e800, 0xc000637040, 0x0)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/internal/cache/cache.go:447 +0x81
k8s.io/kubernetes/pkg/scheduler/internal/cache.(*schedulerCache).updatePod(0xc00034e360, 0xc00056e800, 0xc0007da000, 0x1, 0x2c79be0)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/internal/cache/cache.go:431 +0x35
k8s.io/kubernetes/pkg/scheduler/internal/cache.(*schedulerCache).UpdatePod(0xc00034e360, 0xc00056e800, 0xc0007da000, 0x0, 0x0)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/internal/cache/cache.go:509 +0x224
k8s.io/kubernetes/pkg/scheduler.(*Scheduler).updatePodInCache(0xc0004f83f0, 0x1ac7160, 0xc00056e800, 0x1ac7160, 0xc0007da000)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/eventhandlers.go:255 +0x91
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/controller.go:234
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.FilteringResourceEventHandler.OnUpdate(0xc0002c23c0, 0x1dc06c0, 0xc0003c4840, 0x1ac7160, 0xc00056e800, 0x1ac7160, 0xc0007da000)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/controller.go:269 +0x120
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go:775 +0x1f1
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000979f60)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00050ff60, 0x1d7b660, 0xc0008ca480, 0xc000892001, 0xc0008d4120)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xa3
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000979f60, 0x3b9aca00, 0x0, 0x1, 0xc0008d4120)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0xe2
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(...)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc000211580)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go:771 +0x95
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc000158140, 0xc0008d03a0)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:73 +0x51
created by k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x62
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1647290]

goroutine 247 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x105
panic(0x188b100, 0x2c28480)
	/usr/local/go/src/runtime/panic.go:967 +0x166
k8s.io/kubernetes/pkg/scheduler/types.(*NodeInfo).RemovePod(0xc0005f1450, 0xc00056e800, 0xc000884930, 0x9)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/types/node_info.go:555 +0x610
k8s.io/kubernetes/pkg/scheduler/internal/cache.(*schedulerCache).removePod(0xc00034e360, 0xc00056e800, 0xc000637040, 0x0)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/internal/cache/cache.go:447 +0x81
k8s.io/kubernetes/pkg/scheduler/internal/cache.(*schedulerCache).updatePod(0xc00034e360, 0xc00056e800, 0xc0007da000, 0x1, 0x2c79be0)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/internal/cache/cache.go:431 +0x35
k8s.io/kubernetes/pkg/scheduler/internal/cache.(*schedulerCache).UpdatePod(0xc00034e360, 0xc00056e800, 0xc0007da000, 0x0, 0x0)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/internal/cache/cache.go:509 +0x224
k8s.io/kubernetes/pkg/scheduler.(*Scheduler).updatePodInCache(0xc0004f83f0, 0x1ac7160, 0xc00056e800, 0x1ac7160, 0xc0007da000)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/eventhandlers.go:255 +0x91
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/controller.go:234
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.FilteringResourceEventHandler.OnUpdate(0xc0002c23c0, 0x1dc06c0, 0xc0003c4840, 0x1ac7160, 0xc00056e800, 0x1ac7160, 0xc0007da000)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/controller.go:269 +0x120
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go:775 +0x1f1
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000979f60)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00050ff60, 0x1d7b660, 0xc0008ca480, 0xc000892001, 0xc0008d4120)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xa3
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000979f60, 0x3b9aca00, 0x0, 0x1, 0xc0008d4120)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0xe2
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(...)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc000211580)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go:771 +0x95
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc000158140, 0xc0008d03a0)
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:73 +0x51
created by k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/home/thomas/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x62
```

--- a/pkg/scheduler/internal/cache/cache.go
+++ b/pkg/scheduler/internal/cache/cache.go
@@ -428,6 +428,13 @@ func (cache *schedulerCache) addPod(pod *v1.Pod) {
 
 // Assumes that lock is already acquired.
 func (cache *schedulerCache) updatePod(oldPod, newPod *v1.Pod) error {
+	if _, ok := cache.nodes[newPod.Spec.NodeName]; !ok {
+		// The node might have been deleted already.
+		// This is not a problem in the case where a pod update arrives before the
+		// node creation, because we will always have a create pod event before
+		// that, which will create the placeholder node item.
+		return nil
+	}
 	if err := cache.removePod(oldPod); err != nil {
 		return err
 	}

--- a/pkg/scheduler/internal/cache/cache_test.go
+++ b/pkg/scheduler/internal/cache/cache_test.go
@@ -1232,7 +1232,7 @@ func TestSchedulerCache_UpdateSnapshot(t *testing.T) {
 
 	// Create a few pods for tests.
 	pods := []*v1.Pod{}
-	for i := 0; i < 10; i++ {
+	for i := 0; i < 20; i++ {
 		pod := &v1.Pod{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      fmt.Sprintf("test-pod%v", i),
@@ -1240,7 +1240,7 @@ func TestSchedulerCache_UpdateSnapshot(t *testing.T) {
 				UID:       types.UID(fmt.Sprintf("test-puid%v", i)),
 			},
 			Spec: v1.PodSpec{
-				NodeName: fmt.Sprintf("test-node%v", i),
+				NodeName: fmt.Sprintf("test-node%v", i%10),
 			},
 		}
 		pods = append(pods, pod)
@@ -1309,6 +1309,13 @@ func TestSchedulerCache_UpdateSnapshot(t *testing.T) {
 	addPodWithAffinity := func(i int) operation {
 		return func() {
 			if err := cache.AddPod(podsWithAffinity[i]); err != nil {
+				t.Error(err)
+			}
+		}
+	}
+	removePod := func(i int) operation {
+		return func() {
+			if err := cache.RemovePod(pods[i]); err != nil {
 				t.Error(err)
 			}
 		}
@@ -1382,13 +1389,6 @@ func TestSchedulerCache_UpdateSnapshot(t *testing.T) {
 			expected: []*v1.Node{nodes[6], nodes[5], nodes[2], nodes[0]},
 		},
 		{
-			name: "Remove non-existing node",
-			operations: []operation{
-				addNode(0), addNode(1), updateSnapshot(),
-			},
-			expected: []*v1.Node{nodes[1], nodes[0]},
-		},
-		{
 			name: "Update some nodes",
 			operations: []operation{
 				addNode(0), addNode(1), addNode(5), updateSnapshot(), updateNode(1),
@@ -1444,11 +1444,19 @@ func TestSchedulerCache_UpdateSnapshot(t *testing.T) {
 			expected: []*v1.Node{nodes[0], nodes[4], nodes[2]},
 		},
 		{
-			name: "Remove pod from non-existing node",
+			name: "Add pod before its node",
 			operations: []operation{
-				addNode(0), addPod(0), addNode(2), updateSnapshot(),
+				addNode(0), addPod(1), updatePod(1), addNode(1),
 			},
-			expected: []*v1.Node{nodes[2], nodes[0]},
+			expected: []*v1.Node{nodes[1], nodes[0]},
+		},
+		{
+			name: "Remove node before its pods",
+			operations: []operation{
+				addNode(0), addNode(1), addPod(1), addPod(11),
+				removeNode(1), updatePod(1), updatePod(11), removePod(1), removePod(11),
+			},
+			expected: []*v1.Node{nodes[0]},
 		},
 		{
 			name: "Add Pods with affinity",

